Fix sorting issues with collections.

ojlyytinen opened the pull request:
Fixes #1272. There are a some potential issues with this pull request which should be considered before merging.

As discussed in the issue page, collections don't have date_uploaded or date_modified set so something else needs to be used for sorting. I changed sorting to use create_date and modified_date, which are indexed in Solr as system_create_dtsi and system_modified_dtsi. This change applies to both GenericFiles and Collections. One file that needed changing was the [catalog_controller generator template](https://github.com/projecthydra/sufia/blob/master/lib/generators/sufia/templates/catalog_controller.rb#L22-L28) where it's difficult to differentiate the behaviour between GenericFiles and Collections.

At first I thought that create_date and date_uploaded would be identical in GenericFiles so this wouldn't be a problem, but I then came across [this comment](https://github.com/projecthydra/sufia/blob/master/sufia-models/app/models/concerns/sufia/generic_file/metadata.rb#L51-L55) in metadata.rb

```ruby
# We reserve date_uploaded for the original creation date of the record.
# For example, when migrating data from a fedora3 repo to fedora4,
# fedora's system created date will reflect the date when the record
# was created in fedora4, but the date_uploaded will preserve the
# original creation date from the old repository.
```

So using the create_date instead of date_uploaded is less than ideal in cases where migration has been used at some point. (As a side note, not having the original creation date of collections preserved at all in migrations might be a problem in itself.)

Note that this PR also adds a small improvement that was not part of the original issue but is related. When clicking the Add Files link in a collection page, I set it to pass the id of the collection as a parameter which goes through to the collection list template so that the correct collection can be preselected there. The list used to have the checked attribute set for all options which resulted in the last collection in the list to be selected (at least on the browser I use). The user however might expect the collection where they started the operation at to be selected and if they don't pay attention, they could easily add the files to the wrong collection.

mjgiarlo commented:
I'll review this very soon!  Sorry for the wait.

hectorcorrea commented:
:+1: Nice job @ojlyytinen

ojlyytinen commented:
Does this look ok as it is then, despite the possible issues I mentioned? I'm mainly concerned about using create_date instead of date_uploaded in files where it seems like date_uploaded was specifically made to be preserved through migrations whereas create_date would be reset. So after migration, it would be impossible to sort files by their original creation date.

I can add a couple more tests just to have the coverage check pass.

mjgiarlo commented:
Since you offered, @ojlyytinen, sure, go ahead and toss a couple extra tests in. Because this is a thorough and well documented PR, I haven't had a chance to fully review it yet -- apologies that you've had to wait for a merge.  You have raised important issues in your comments on this PR that we'll want to consider and discuss here.  Hoping some other folks will chime in. @cam156 @jcoyne @awead ?

awead commented on spec/features/collection_spec.rb:
```diff
@@ -140,6 +140,82 @@ def create_collection(title, description)
+  describe 'add files to collection' do
+    let!(:collection1) do
+      Collection.create(title: title1, description: description1,
+                        members: []) { |c| c.apply_depositor_metadata(user.user_key) }
```
This is the same as collection1 above. Can you just reuse the existing one?

awead commented on spec/features/collection_spec.rb:
```diff
@@ -140,6 +140,82 @@ def create_collection(title, description)
+    end
+    let!(:collection2) do
+      Collection.create(title: title2, description: description2,
+                        members: []) { |c| c.apply_depositor_metadata(user.user_key) }
```
dito here. Reuse existing collection2?

awead commented on spec/views/collections/_form_for_select_collection.html.erb_spec.rb:
```diff
@@ -0,0 +1,49 @@
+  end
+
+  it "sorts the collections" do
+    render
```
try assigning render to a `let!` variable so it might only render the page once.

awead commented on app/models/concerns/sufia/solr_document_behavior.rb:
```diff
@@ -50,6 +50,16 @@ def date_uploaded
+      return unless field.present?
+      begin
+        Date.parse(field).to_formatted_s(:standard)
+      rescue
```
It might be better to rescue from the specific data parsing error, instead of just any error.

awead commented on app/controllers/concerns/sufia/my_controller_behavior.rb:
```diff
@@ -52,6 +52,8 @@ def index
       @batch_size_on_other_page = batch_size - count_on_page
       @batch_part_on_other_page = (@batch_size_on_other_page) > 0
 
+      @add_files_to_collection = params[:add_files_to_collection] || ''
```
try using fetch here:
```
@add_files_to_collection = fetch(params[:add_files_to_collection], '')
```

mjgiarlo replied:
Typo? I think you meant:

```ruby
@add_files_to_collection = params.fetch(:add_files_to_collection, '')
```

awead replied:
sorry, pre :coffee: comment.

mjgiarlo replied:
happens!

awead commented on app/views/collections/_form_for_select_collection.html.erb:
```diff
@@ -13,9 +13,10 @@
                 <legend><%= t("sufia.collection.select_form.select_heading") %></legend>
               <ul>
-                <% user_collections.sort { |c1,c2| c1['date_modified_dtsi'] <=> c2['date_modified_dtsi'] }.each do |collection|  %>
+                <% user_collections.sort { |c1,c2| c2['system_create_dtsi'] <=> c1['system_create_dtsi'] }.each_with_index do |collection,index|  %>
```
using solr_name or `CatalogController.uploaded_field` here would avoid having to know _dtsi.

awead commented on app/models/concerns/sufia/solr_document_behavior.rb:
```diff
@@ -50,6 +50,16 @@ def date_uploaded
     end
 
+    def create_date
+      field = self[Solrizer.solr_name("system_create", :stored_sortable, type: :date)]
```
you could call `CatalogController.uploaded_field` instead.

awead commented on spec/models/solr_document_spec.rb:
```diff
@@ -10,6 +10,15 @@
+      expect(subject.create_date).to eq '03/14/2013'
+    end
+  end
+
```
might be good to test your rescue clause here too.

awead commented:
Looks good overall, @ojlyytinen. I added a few comments to shore things up a bit.

mjgiarlo commented:
@awead @cam156 @hectorcorrea Are we concerned that this changes the date for sorting? (Using the F4 date rather than date_uploaded, which we used to carry forward F3 dates.)  If not, I'm :+1:.

awead commented:
@mjgiarlo if we always refer to the sorting field as `CatalogController.uploaded_field` then it ought be easy to change around.

mjgiarlo commented:
@awead so we'd deal with this in ScholarSphere (which this PR will break because we need the dates that reflect F3 values) by redefining `CatalogController.uploaded_field`?

awead commented:
@mjgiarlo well, since we've defined CatalogController.uploaded_field, Sufia would use whatever that field is.

mjgiarlo commented:
@awead I'm not following.  Once @ojlyytinen makes the requested tweaks (including using `CatalogController.uploaded_field`, if we merge this, you're saying this PR won't break ScholarSphere because we already have this in place for our F3 date needs? https://github.com/psu-stewardship/scholarsphere/blob/develop/app/controllers/catalog_controller.rb#L21-L27

If so, :+1:

awead commented:
@mjgiarlo it shouldn't, because Sufia will look to our controller's class variable to determine what the field is. At least that's how I'm thinking it will.

ojlyytinen commented:
I've made the changes, with one exception. I'm not quite sure what @awead meant with putting the render in a let! variable. I thought that let! variables get evaluated once per example anyway so it still wouldn't render the page only once. And besides, the middle example assigns one extra variable and needs its own rendering anyway. Or am I completely misunderstanding what you meant?

For the duplicate collection definitions, I moved them up a level so they don't need to be repeated. I also changed them to lazy evaluation so they won't be created for all examples. Their creation is then forced where needed by just referencing them.

All system_create_dtsi's should be changed to CatalogController.uploaded_field now. There were a few more in views/my/_index_partials/_default_group.html.erb

awead commented:
:+1:

ojlyytinen commented:
Pushed in one more update with a few extra examples in solr_document_spec to satisfy the coverage check.

mjgiarlo commented:
Thanks, @ojlyytinen, for the contribution and for your patience!
